src/conmon.c

```diff
@@ -728,6 +701,91 @@ static gboolean conn_sock_cb(int fd, GIOCondition condition, gpointer user_data)
 	return G_SOURCE_REMOVE;
 }
 
+static gboolean conn_sock_cb(G_GNUC_UNUSED int fd, GIOCondition condition, gpointer user_data);
+
+static void sock_try_write_to_masterfd_stdin(struct conn_sock_s *sock)
+{
+	if (sock->remaining) {
+		ssize_t w;
+
+		w = write(masterfd_stdin, sock->buf + sock->off, sock->remaining);
+		if (w < 0) {
+			nwarn("Failed to write to container stdin");
+		} else {
+			sock->off += w;
+			sock->remaining -= w;
+		}
+	}
+}
+
+static void write_to_masterfd_stdin(gpointer data, gpointer user_data)
+{
+	struct conn_sock_s *sock = (struct conn_sock_s *)data;
+	bool *has_data = user_data;
+
+	sock_try_write_to_masterfd_stdin(sock);
+
+	if (sock->remaining)
+		*has_data = TRUE;
+	else if (sock->data_ready) {
+		sock->data_ready = false;
+		g_unix_fd_add(sock->fd, G_IO_IN | G_IO_HUP | G_IO_ERR, conn_sock_cb, sock);
+	}
+}
+
+static gboolean masterfd_write_cb(G_GNUC_UNUSED int fd, G_GNUC_UNUSED GIOCondition condition, G_GNUC_UNUSED gpointer user_data)
+{
+	bool has_data = FALSE;
+
+	if (masterfd_stdin < 0)
+		return G_SOURCE_REMOVE;
+
+	g_ptr_array_foreach(conn_socks, write_to_masterfd_stdin, &has_data);
+	if (has_data)
+		return G_SOURCE_CONTINUE;
+	return G_SOURCE_REMOVE;
+}
+
+static gboolean read_conn_sock(struct conn_sock_s *sock)
+{
+	ssize_t num_read;
+
+	/* There is still data in the buffer.  */
+	if (sock->remaining) {
+		sock->data_ready = true;
+		return G_SOURCE_REMOVE;
+	}
+
+	num_read = read(sock->fd, sock->buf, CONN_SOCK_BUF_SIZE);
+	if (num_read < 0)
+		return G_SOURCE_CONTINUE;
+
+	if (num_read > 0) {
+		sock->remaining = num_read;
+		sock->off = 0;
+
+		sock_try_write_to_masterfd_stdin(sock);
+
+		/* Not everything was written to stdin, let's wait the fd is ready.  */
+		if (sock->remaining)
+			g_unix_fd_add(masterfd_stdin, G_IO_OUT, masterfd_write_cb, NULL);
+	} else {
```

Comment:
you could do `if (num_read == 0) {return ...}` above the `num_read > 0` block, which could then be indented in